### PR TITLE
fix(workers/pr): improve deduplication in updates-table

### LINKED ISSUE #26073: Dependency Duplicated In Table In Pull Request Description
### Discussed in https://github.com/renovatebot/renovate/discussions/26025

<div type='discussions-op-text'>

<sup>Originally posted by **Kurt-von-Laven** November 29, 2023</sup>
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

We are using Forking Renovate. The description of ScribeMD/docker-cache#729 lists Yarn v2+ twice, once with a link to [a site that never loads](https://app.renovatebot.com/package-diff?name=yarn&from=3.6.4&to=4.0.2) and once without. The table in the commit message body correctly lists `@yarnpkg/cli` once. I am guessing the duplicate in the PR description stems from the fact that both the npm manager and our regex manager bump Yarn. It makes me wonder if our regex manager may be configured incorrectly in some subtle way. How are rows in the table in the PR description deduplicated?

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Improved the de-duplication among rows of the update pr table by only using the necessary properties to compare the upgrades rather than comparing the entire row (which is a string)
The properties used for comparison are: `depName`, `depType`, `updateType`, `currentVersion/Value` and `newVersion/Value`

## Context
- Closes: #26073

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
Test Run: https://github.com/RahulGautamSingh-testing/26073/pull/2

File: lib/workers/repository/update/pr/body/updates-table.ts [hunk 1]
```diff
@@ -44,7 +44,10 @@ export function getPrUpdatesTable(config: BranchConfig): string {
     logger.warn('getPrUpdatesTable - prBodyColumns is undefined');
     return '';
   }
-  const tableValues = config.upgrades
+
+  // filter duplicate upgrades
+  const uniqueUpgrades = filterDuplicateUpgrades(config.upgrades);
+  const tableValues = uniqueUpgrades
```
Comment (viceice) [inline lib/workers/repository/update/pr/body/updates-table.ts:50]:
```suggestion
  const tableValues = filterDuplicateUpgrades(config.upgrades)
```

**Review by viceice — COMMENTED** [review]
missing coverage

**Review by viceice — COMMENTED** [review]
🤔 this can cause dropping rows with more information. 

so for the reproduction sample, if the regex manager would come first, the compare link would be gone.

@rarkins I think this needs to be feature flagged. I don't want to filter those in my PR.

maybe we need to merge those rows instead of simply dropping.

**Comment by rarkins** [comment]
What if instead of this, we add logic to _enrich_ rows which are "duplicates" but lacking in metadata? Then they will deduplicate themselves in a later step. Either that or picking the row which has most metadata.

I don't want to feature flag this - it's too trivial to deserve that

**Review by viceice — CHANGES_REQUESTED** [review]
needs rework

**Comment by RahulGautamSingh** [comment]
I don't think any other combination of managers will create duplicate deps unless one of them is the regex manager. Therefore, we can just filter regex manager's row in cases of duplicate rows.

**Review by viceice — COMMENTED** [review]
as discussed this still causes bad usability, eg npm and nodeenv manager

**Review by viceice — APPROVED** [review]

**Pull request merged**